src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.State.cs

```diff
@@ -0,0 +1,297 @@
+                // then it's an actual linked file (i.e. not a MTFM project). We don't support that, because 
+                // we don't know which default namespace and folder path we should use to construct target
+                // namespace.
+                if (linkedDocumentids.Any(id => 
+                !PathUtilities.PathsEqual(solution.GetDocument(id).Project.FilePath, document.Project.FilePath)))
+                {
+                    allDocumentIds = default;
+                    return false;
+                }
+
+                // Now determine if the actual file path matches its logical path in project 
+                // which is constructed as <project root path>\Logical\Folders\. The refactoring 
+                // is triggered only when the two match. The reason of doing this is we don't really know
+                // the user's intention of keeping the file path out-of-sync with its logical path.
+                var projectRoot = PathUtilities.GetDirectoryName(document.Project.FilePath);
+                var folderPath = Path.Combine(document.Folders.ToArray());
+
+                var absoluteDircetoryPath = PathUtilities.GetDirectoryName(document.FilePath);
+                var logicalDirectoryPath = PathUtilities.CombineAbsoluteAndRelativePaths(projectRoot, folderPath);
+
+                if (PathUtilities.PathsEqual(absoluteDircetoryPath, logicalDirectoryPath))
+                {
+                    allDocumentIds = linkedDocumentids.Add(document.Id);
+                    return true;
+                }
+                else
+                {
+                    allDocumentIds = default;
+                    return false;
+                }
+            }
+
+            private async static Task<(bool shouldTriggerRefactoring, string defaultNamespace, string declaredNamespace)> 
+                TryGetDefaultNamespaceAndNamespaceDeclarationAsync(
+                ImmutableArray<Document> documents,
+                AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax> service,
+                ISyntaxFactsService syntaxFacts, 
+                TextSpan textSpan, 
+                CancellationToken cancellationToken)
+            {
+                // For all projects containing all the linked documents, bail if 
+                // 1. Any of them doesn't have default namespace, or
+                // 2. Multiple default namespace are found. (this might be possible by tweaking project file).
+                // The refactoring depends on a single default namespace to operate.
+                var defaultNamespaceFromProjects = new HashSet<string>(syntaxFacts.StringComparer);
+                if (documents.Any(doc =>
+                {
+                    var defaultNS = doc.Project.DefaultNamespace;
+                    defaultNamespaceFromProjects.Add(defaultNS);
+                    return defaultNS == null;
+                }))
```

Comment:
this is super confusing.  please just extract a local and test hte local.  inlining like this just hurts readability.

Reply:
also, mutating objects in a .Any call is super unpleasant.  please don't do that.